File: backend/storage-service/src/Fix.Storage.Domain/AggregateRoots/Files/Entities/File.cs
namespace Fix.Storage.Domain.AggregateRoots.Files;

/// <summary>
/// Arquivo enviado por um usuário de uma organização: onde está (S3), o tipo e os contadores do processamento. As linhas
/// ficam em <see cref="FileLine"/>. Status e contadores são alterados de forma atômica pelo repositório (várias linhas
/// podem terminar ao mesmo tempo, em consumidores diferentes).
/// </summary>
public sealed class File : AggregateRoot
{
    private File()
    {
    }

    public Guid OrganizationId { get; private set; }

    public Guid UploadedBy { get; private set; }

    public FileKind Kind { get; private set; }

    public string FileName { get; private set; } = string.Empty;

    public string ContentType { get; private set; } = string.Empty;

    public long SizeBytes { get; private set; }

    /// <summary>Chave do objeto no bucket.</summary>
    public string StorageKey { get; private set; } = string.Empty;

    /// <summary>s3://bucket/chave</summary>
    public string StorageUrl { get; private set; } = string.Empty;

    public FileStatus Status { get; private set; }

    public DateTimeOffset UploadedAt { get; private set; }

    public DateTimeOffset? FinishedAt { get; private set; }

    public int TotalLines { get; private set; }

    public int ProcessedLines { get; private set; }

    public int SucceededLines { get; private set; }

    public int FailedLines { get; private set; }

    /// <summary>Falha do arquivo inteiro (ex.: não pôde ser lido).</summary>
    public string? Error { get; private set; }

    public bool IsFinished => Status is FileStatus.Completed or FileStatus.CompletedWithErrors or FileStatus.Failed or FileStatus.Stored;

    public int ProgressPercent => TotalLines == 0 ? (IsFinished ? 100 : 0) : (int)Math.Floor(ProcessedLines * 100m / TotalLines);

    /// <summary>Recebido: os tipos processados vão para a fila de leitura; os demais já terminam armazenados.</summary>
    public static File Receive(
        Guid organizationId,
        Guid uploadedBy,
        FileKind kind,
        string fileName,
        string contentType,
        long sizeBytes,
        string storageKey,
        string storageUrl,
        DateTimeOffset now) => new()
    {
        OrganizationId = organizationId,
        UploadedBy = uploadedBy,
        Kind = kind,
        FileName = fileName,
        ContentType = contentType,
        SizeBytes = sizeBytes,
        StorageKey = storageKey,
        StorageUrl = storageUrl,
        UploadedAt = now,
        Status = kind.IsProcessed() ? FileStatus.Received : FileStatus.Stored,
        FinishedAt = kind.IsProcessed() ? null : now,
    };

    /// <summary>Linhas lidas: começa o processamento. Falso se o arquivo já saiu de Received (reentrega do evento).</summary>
    public bool StartProcessing(int totalLines)
    {
        if (Status != FileStatus.Received)
        {
            return false;
        }

        Status = FileStatus.Processing;
        TotalLines = totalLines;
        return true;
    }

    public void MarkFailed(string error, DateTimeOffset at)
    {
        Status = FileStatus.Failed;
        Error = error;
        FinishedAt = at;
    }

    /// <summary>Uma linha terminou; a última fecha o arquivo com o status final.</summary>
    public void RegisterLineResult(bool succeeded, DateTimeOffset at)
    {
        ProcessedLines++;
        if (succeeded) SucceededLines++;
        else FailedLines++;

        if (Status == FileStatus.Processing && ProcessedLines >= TotalLines)
        {
            Status = FinalStatus(FailedLines);
            FinishedAt = at;
        }
    }

    /// <summary>Status final a partir das linhas que falharam.</summary>
    public static FileStatus FinalStatus(int failedLines) => failedLines > 0 ? FileStatus.CompletedWithErrors : FileStatus.Completed;
}

File: frontend/web/src/domain/file.ts
/** Tipos de upload: usuários, políticas, mandatos e boletas são processados linha a linha (só criação); documentos só armazenados. */
export const FILE_KINDS = ['Users', 'Policies', 'Mandates', 'Orders', 'Documents'] as const;
export type FileKind = (typeof FILE_KINDS)[number];

export const FILE_STATUSES = ['Received', 'Processing', 'Completed', 'CompletedWithErrors', 'Failed', 'Stored'] as const;
export type FileStatus = (typeof FILE_STATUSES)[number];

export const FILE_LINE_STATUSES = ['Pending', 'Succeeded', 'Failed'] as const;
export type FileLineStatus = (typeof FILE_LINE_STATUSES)[number];

/** Limite por arquivo (o mesmo do storage-service). */
export const MAX_UPLOAD_BYTES = 20 * 1024 * 1024;

/** Tipos cujas linhas viram cadastros no core. */
export const PROCESSED_KINDS: readonly FileKind[] = ['Users', 'Policies', 'Mandates', 'Orders'];

/** Trecho da URL de cada tipo (/uploads/mandatos). */
export const fileKindSlug: Record<FileKind, string> = {
  Users: 'usuarios',
  Policies: 'politicas',
  Mandates: 'mandatos',
  Orders: 'boletas',
  Documents: 'documentos',
};

export const fileKindFromSlug = (slug: string | undefined): FileKind | null =>
  FILE_KINDS.find((kind) => fileKindSlug[kind] === slug) ?? null;

/** Arquivo em andamento: ainda recebe eventos de progresso. */
export const isFileRunning = (status: FileStatus): boolean => status === 'Received' || status === 'Processing';

export interface UploadedFile {
  id: string;
  kind: FileKind;
  status: FileStatus;
  fileName: string;
  contentType: string;
  sizeBytes: number;
  organizationId: string;
  uploadedBy: string;
  uploadedAt: string;
  finishedAt: string | null;
  totalLines: number;
  processedLines: number;
  succeededLines: number;
  failedLines: number;
  /** Falha do arquivo inteiro (ex.: não pôde ser lido). */
  error: string | null;
  /** 0–100 */
  progressPercent: number;
}

export interface FileLine {
  id: string;
  /** 1 = primeira linha de dados (o cabeçalho não conta). */
  number: number;
  status: FileLineStatus;
  /** Coluna → valor, como veio no arquivo. */
  values: Record<string, string>;
  /** Erro da linha ou resumo do que foi feito. */
  message: string | null;
  /** Código do que foi criado (MD-07, HX-0012). */
  resultCode: string | null;
  processedAt: string | null;
}

/** Card de um tipo na tela de uploads. */
export interface FileKindSummary {
  kind: FileKind;
  total: number;
  processing: number;
  withErrors: number;
  lastUploadedAt: string | null;
  /** O usuário tem a regra exigida para enviar este tipo. */
  canUpload: boolean;
}

export interface FileTemplateColumn {
  name: string;
  required: boolean;
  description: string;
  example: string;
}

export interface FileTemplate {
  kind: FileKind;
  /** false = só armazena. */
  processed: boolean;
  /** Vazio = qualquer extensão. */
  acceptedExtensions: string[];
  columns: FileTemplateColumn[];
}

/** Evento de progresso recebido ao vivo (SSE). */
export interface FileProgress {
  fileId: string;
  organizationId: string;
  userId: string;
  kind: FileKind;
  fileName: string;
  status: FileStatus;
  totalLines: number;
  processedLines: number;
  succeededLines: number;
  failedLines: number;
  percent: number;
  error: string | null;
  /** A linha que acabou de terminar (ausente nos eventos do arquivo inteiro). */
  line: { number: number; status: FileLineStatus; message: string | null; resultCode: string | null } | null;
}

/** Aplica um evento de progresso ao arquivo carregado. */
export const applyProgress = (file: UploadedFile, event: FileProgress): UploadedFile => ({
  ...file,
  status: event.status,
  totalLines: event.totalLines,
  processedLines: event.processedLines,
  succeededLines: event.succeededLines,
  failedLines: event.failedLines,
  progressPercent: event.percent,
  error: event.error,
});
